File: 1113_function_procedure/04_faktorisasi_prima/practice.pas
program factorate;

var
prim:array [0..1000000] of longint;
outp:array [0..1000000] of longint;
n,x,it,y,l:longint;
i:longint;
tr:boolean;


//MENCARI BILANGAN PRIMA 2 - (N DIV 2)
procedure searchPrime(n:longint);
var
i,j:longint;
frek:longint;
begin
	x := 1;
	for i := 2 to (n div 2) do
	begin
		while n mod i = 0 do
		begin
			n := n div i;
			prim[x] := i;
			inc(x);
		end;
	end;
end;

//MENCARI FAKTORASI MENGGUNAKAN REKURSIF, MEMASUKKAN HASIL KE DALAM ARRAY OUTP[]
procedure factorate(n:longint);
var
x,z:longint;
begin
	x := 1;
	if (n = 1) then
	begin
		exit;
	end else
	begin
		while (n mod prim[x] <> 0) do
		begin
			x := x + 1;
		end;
		z := n div prim[x];
		outp[y] := prim[x];
		y := y + 1;
		factorate(z);
	end;
end;

//MENGECEK APAKAH N BILANGAN PRIMA ATAU BUKAN
procedure cekPrim(n:longint);
var
i,frek:longint;
begin
	frek := 0;
	for i := 2 to trunc(sqrt(n)) do
	begin
		if (n mod i = 0) then
		begin
			frek := frek + 1;
		end;
	end;
	if (frek = 0) then
	begin
		writeln (n);
		tr := true;
	end;
end;

begin
	tr := false;
	y := 1;
	readln (n);
	l := n;
	cekPrim(n);
	if (tr = true) then
	begin
		exit;
	end;
	searchPrime(n);
	factorate(n);
	it := 1;
	//OUTPUT
	for i := 1 to y-2 do
	begin
		if (outp[i] = outp[i+1]) then
		begin
			it := it + 1;
		end else if (outp[i] <> outp[i+1]) then
		begin
			if (it = 1) then
			begin
				write (outp[i],' ','x',' ');
				it := 1;
			end else
			begin
				write (outp[i],'^',it,' ','x',' ');
				it := 1;
			end;
		end;	
	end;
	if (it = 1) then
	begin
		writeln (outp[i+1]);
	end else
	begin
		writeln (outp[i+1],'^',it);
	end;
end.

File: 1113_function_procedure/04_faktorisasi_prima/practice.cpp
#include <iostream>
#include <math.h>
#include <stdlib.h>

using namespace std;

const long limit = 1000000;
long prim[limit], outp[limit];
long n, x, it, y, l = 0;
long i = 0;
bool tr;

void searchPrime(long n)
{
  long i, j;
  long frek;
  x = 0;
  for (i = 0; i < n / 2; i++)
  {
    frek = 0;
    if (i > 1)
    {
        for (j = 2; j <= (trunc(sqrt(i))); j++)
        {
          if (i % j == 0)
          {
            frek++;
          }
        }
        if (frek == 0)
        {
          prim[x] = i;
          x++;
        }
    }
  }
}

void showPrime(long n)
{
  cout << "Display PRIME" << endl;
  for (int i = 0; i < n; i++)
  {
    cout << prim[i];
    if (i < n - 1) cout << " ";
  }
  cout << endl;
}

void factorate(long n)
{
  long x, z;
  x = 0;
  if (n == 1)
  {
    // exit(EXIT_FAILURE);
  } else
  {
    // cout << "=== Factorate ===" << endl;
    while (n % prim[x] != 0)
    {
      // cout << "prim[" << x << "]: " << prim[x] << " ";
      x++;
      // cout << "x: " << x << ", ";
    }
    // cout << "; ";
    z = n / prim[x];
    outp[y] = prim[x];
    // cout << "z: " << z << "; " 
    // << "prim[" << x << "]: " << prim[x] << "; " 
    // << "outp[" << y << "]: " << outp[y] << endl;
    y++;
    factorate(z);
  }
}

void showFactor(long n)
{
  cout << "Display FACTOR" << endl;
  for (int i = 0; i < n; i++)
  {
    cout << outp[i];
    if (i < n - 1) cout << " ";
  }
  cout << endl;
}

int main()
{
  tr = false;
	y = 0;
	cin >> n;
	l = n;
	searchPrime(n);
  // showPrime(x);
	factorate(n);
  // showFactor(y);
	it = 1;

	for (i = 0; i < y - 1; i++)
	{
    // cout << "= OUTPUT INNER LOOP =";
		if (outp[i] == outp[i+1])
		{
			it++;
		} else if (outp[i] != outp[i+1])
		{
			if (it == 1)
			{
				cout << outp[i] << " x ";
				it = 1;
			} else
			{
				cout << outp[i] << '^' << it << " x ";
				it = 1;
			}
		}
    // cout << "outp[" << i << "]: " << outp[i] << "; " << endl;
    // if (i < y - 1) cout << " x "
	}

	if (it == 1)
	{
		cout << outp[i] << endl;
	} else
	{
		cout << outp[i] << '^' << it << endl;
	}

  return 0;
}